salt/_modules/metalk8s_volumes.py

```diff
@@ -0,0 +1,360 @@
+    """
+    return _get_volume(name).initialize()
+
+
+def is_formatted(name):
+    """Check if the given volume is formatted.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        bool: True if the volume is already formatted, otherwise False
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.volume_is_formatted example-volume
+    """
+    return _get_volume(name).is_formatted
+
+
+def format(name):
+    """Format the given volume.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        None
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.format example-volume
+    """
+    _get_volume(name).format()
+
+
+# Volume {{{
+
+
+class Volume(object):
+    """Volume interface."""
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, volume):
+        self._volume = volume
+
+    @abc.abstractproperty
+    def exists(self):
+        """Does the backing storage device exists?"""
+        return
+
+    @abc.abstractmethod
+    def create(self):
+        """Create the backing storage device."""
+        return
+
+    @abc.abstractproperty
+    def is_initialized(self):
+        """Check if the backing storage device is initialized."""
+        return
+
+    @abc.abstractmethod
+    def initialize(self):
+        """Initialize the backing storage device."""
+        return
+
+    @abc.abstractproperty
+    def block_device(self):
+        """Path to the backing block device."""
+        return
+
+    @property
+    def is_formatted(self):
+        """Check if the volume is already formatted."""
+        fs_type = self['spec.storageClassName.parameters.fsType']
+        return __salt__['disk.fstype'](self.block_device) == fs_type
+
+    def format(self):
+        """Format the volume.
+
+        The volume is formatted according to its StorageClass.
+        """
+        params = self['spec.storageClassName.parameters']
+        # Options are stored as JSON-encoded list.
+        mkfs_options = json.loads(params['mkfsOptions'])
+        command = ['mkfs']
+        command.extend(['-t', params['fsType']])
+        command.extend(['-U', self['metadata.uid']])
+        command.extend(mkfs_options)
+        command.append(self.block_device)
+        _run_cmd(' '.join(command))
+
+    def __getitem__(self, path):
+        """Return the Volume attribute `attr` from the Volume dict."""
+        return functools.reduce(operator.getitem, path.split('.'), self._volume)
+
+
+# }}}
+# SparseLoopDevice {{{
+
+
+class SparseLoopDevice(Volume):
+    @property
+    def sparse_file(self):
+        return '/var/lib/metalk8s/storage/sparse/{}'.format(
+            self['metadata.name']
+        )
+
+    @property
+    def size(self):
+        return _quantity_to_bytes(self['spec.sparseLoopDevice.size'])
+
+    @property
+    def exists(self):
+        path = self.sparse_file
+        return os.path.isfile(path) and os.path.getsize(path) == self.size
```

Comment:
Why the size check for an existence check? What happens if the size doesn't match (for whatever reason)?

Reply:
If the size doesn't match that means someone other than us created the file, so I wanted to play it safe and don't touch the file.